third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,595 @@
+//!
+//!     if ... goto L1 else ...
+//!     =>
+//!     if ... goto L2 else ...
+//!
+//!     if ... goto ... else L1
+//!     =>
+//!     if ... goto ... else L2
+//!
+//! 2. Removes edges where the source has only one successor and the target has only one predecessor:
+//!     If there is block
+//!     BB1: xx
+//!          goto L2
+//!     and block
+//!
+//!     BB2: L2
+//!          yy
+//!
+//!     which has only one predecessor BB1.
+//!     Then we move BB2 to the end of BB1 so that BB1 becomes
+//!
+//!     BB1: xx
+//!          yy
+//!
+//!     and BB2 is removed.
+//!
+//! 3. Removes all jumps to the next instruction:
+//!     Whenever we have
+//!
+//!     goto L;
+//!     L;
+//!
+//!     replace it by simply
+//!
+//!     L;
+//!
+//! 4. Replaces branch to same target by jump:
+//!     Whenever we have
+//!
+//!     if ... goto L else goto L
+//!
+//!     replace it by simply
+//!
+//!     goto L
+//!
+//! Side effects: remove all annotations.
+
+use itertools::Itertools;
+use move_model::model::FunctionEnv;
+use move_stackless_bytecode::{
+    function_target::FunctionData,
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{Bytecode, Label},
+    stackless_control_flow_graph::{BlockContent, BlockId, DFSLeft, StacklessControlFlowGraph},
+};
+use std::collections::BTreeMap;
+
+/// Simplifies the control flow graph as described in the module doc
+pub struct ControlFlowGraphSimplifier {}
+
+impl FunctionTargetProcessor for ControlFlowGraphSimplifier {
+    /// Does the first four transformations described in the module doc
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let mut transformer = ControlFlowGraphSimplifierTransformation::new(data);
+        transformer.transform();
+        transformer.data
+    }
+
+    fn name(&self) -> String {
+        "ControlFlowGraphSimplifier".to_owned()
+    }
+}
+
+/// State of the transformation
+struct ControlFlowGraphSimplifierTransformation {
+    data: FunctionData,
+}
+
+impl ControlFlowGraphSimplifierTransformation {
+    pub fn new(data: FunctionData) -> Self {
+        Self { data }
+    }
+
+    /// Does the first four transformations described in the module doc
+    fn transform(&mut self) {
+        self.eliminate_branch_to_same_target();
+        let cfg_code_generator =
+            ControlFlowGraphCodeGenerator::new(std::mem::take(&mut self.data.code));
+        let mut empty_blocks_remover = EmptyBlockRemover::new(cfg_code_generator);
+        empty_blocks_remover.transform();
+        let mut redundant_jump_remover = RedundantJumpRemover::new(empty_blocks_remover.0);
+        redundant_jump_remover.transform();
+        self.data.code = redundant_jump_remover.0.gen_code(true);
```

Comment:
This code would be clearer if you used more abstraction.  These single-element tuple classes are just holders for argument and result for a single transform routine.  How about instead expose the transform routines as functions that operate on CFGCG or bytecode vector directly?

Your current code is identical (modulo renaming, rephrasing, and possible typos) to:
```
        self.eliminate_branch_to_same_target();
        let cfg_code_generator =
            ControlFlowGraphCodeGenerator::new(std::mem::take(&mut self.data.code));

        let mut empty_blocks_remover = EmptyBlockRemover::new(cfg_code_generator);
        empty_blocks_remover.transform();
        let res1 = empty_blocks_remover.0;

        let mut redundant_jump_remover = RedundantJumpRemover::new(res1);
        redundant_jump_remover.transform();
        let res2 = redundant_jump_remover.0;
        let code2 = res2.gen_code(true);

        // may introduce new empty blocks
        let cfgcg2 = ControlFlowGraphCodeGenerator::new(code2);
        let mut empty_blocks_remover = EmptyBlockRemover::new(cfgcg2);
        empty_blocks_remover.transform();
        let res3 = empty_blocks_remover.0;

        self.data.code = res3.gen_code(true);
        self.eliminate_branch_to_same_target();
        self.data.annotations.clear()
```
If we replace the various single-element tuple structs above with single functions of the current struct (`Self`), then we can write this more simply.  I leave the intermediate values in to keep the structure identical here:
```
        let code = std::mem::take(&mut self.data.code);
        let code2 = Self::eliminate_branch_to_same_target(code1);
        let cfg_cg = ControlFlowGraphCodeGenerator::new(code2);

        let cfg_cg1 = Self::remove_empty_blocks(cfg_cg);
        
        let cfg_cg2 = Self::remove_redundant_jumps(cfg_cg1);
        let code2 = cfg_cg2.gen_code(true);

        // may introduce new empty blocks
        let cfg_cg3 = ControlFlowGraphCodeGenerator::new(code2);
        let cfg_cg4 = Self::remove_empty_blocks(cfg_cg3);
        
        let code4 = cfg_cg4.gen_code(true)
        let code5 = Self::eliminate_branch_to_same_target(code4);
        self.data_code = code5;
        self.data.annotations.clear()
```

Reply:
done